File: dotnet/tests/Obscura.Cdp.Tests/ExecutionContextPrunedOnNavigation.cs
using System.Text.Json.Nodes;

using Xunit;

namespace Obscura.Cdp.Tests;

/// <summary>
/// The xUnit port of <c>crates/obscura-cdp/tests/execution_context_pruned_on_navigation.rs</c>.
/// </summary>
/// <remarks>
/// Regression for issue #407: navigation must retire a page's old execution context ids
/// without deleting live contexts owned by another page.
/// </remarks>
[Collection(CdpDomainCollection.Name)]
public sealed class ExecutionContextPrunedOnNavigation
{
    [Fact]
    public async Task NavigationPrunesOnlyTheNavigatedPagesStaleContextIds()
    {
        var ctx = CdpContext.New();
        string firstPage = ctx.CreatePage();
        string secondPage = ctx.CreatePage();
        ctx.Sessions["first"] = firstPage;
        ctx.Sessions["second"] = secondPage;

        long stale = (await CoreCdp.CdpAsync(
            ctx, 1, "Page.createIsolatedWorld",
            new JsonObject { ["worldName"] = "first-world" }, "first"))["executionContextId"]
            .AsI64()!.Value;
        long otherPage = (await CoreCdp.CdpAsync(
            ctx, 2, "Page.createIsolatedWorld",
            new JsonObject { ["worldName"] = "second-world" }, "second"))["executionContextId"]
            .AsI64()!.Value;

        await CoreCdp.DispatchAsync(
            ctx,
            3,
            "Page.navigate",
            new JsonObject
            {
                ["url"] = "data:text/html,<p>replacement</p>",
                ["waitUntil"] = "load",
            },
            "first");

        CdpResponse staleResult = await CoreCdp.DispatchAsync(
            ctx, 4, "Runtime.evaluate",
            new JsonObject { ["expression"] = "1", ["contextId"] = stale }, "first");
        Assert.NotNull(staleResult.Error);
        Assert.Contains("Cannot find context", staleResult.Error!.Message, StringComparison.Ordinal);

        CdpResponse otherResult = await CoreCdp.DispatchAsync(
            ctx, 5, "Runtime.evaluate",
            new JsonObject { ["expression"] = "1", ["contextId"] = otherPage }, "second");
        Assert.True(otherResult.Error is null, "other page context was pruned");
    }
}

File: .reference/obscura/crates/obscura-cdp/tests/execution_context_pruned_on_navigation.rs
// Regression for issue #407: navigation must retire a page's old execution
// context ids without deleting live contexts owned by another page.

use obscura_cdp::dispatch::{dispatch, CdpContext};
use obscura_cdp::types::CdpRequest;
use serde_json::{json, Value};

async fn cdp(
    ctx: &mut CdpContext,
    id: u64,
    method: &str,
    params: Value,
    session_id: &str,
) -> obscura_cdp::types::CdpResponse {
    dispatch(&CdpRequest {
        id,
        method: method.to_string(),
        params,
        session_id: Some(session_id.to_string()),
    }, ctx).await
}

#[tokio::test(flavor = "current_thread")]
async fn navigation_prunes_only_the_navigated_pages_stale_context_ids() {
    let mut ctx = CdpContext::new();
    let first_page = ctx.create_page();
    let second_page = ctx.create_page();
    ctx.sessions.insert("first".to_string(), first_page);
    ctx.sessions.insert("second".to_string(), second_page);

    let stale = cdp(
        &mut ctx, 1, "Page.createIsolatedWorld",
        json!({"worldName": "first-world"}), "first",
    ).await.result.unwrap()["executionContextId"].as_i64().unwrap();
    let other_page = cdp(
        &mut ctx, 2, "Page.createIsolatedWorld",
        json!({"worldName": "second-world"}), "second",
    ).await.result.unwrap()["executionContextId"].as_i64().unwrap();

    cdp(
        &mut ctx, 3, "Page.navigate",
        json!({"url": "data:text/html,<p>replacement</p>", "waitUntil": "load"}),
        "first",
    ).await;

    let stale_result = cdp(
        &mut ctx, 4, "Runtime.evaluate",
        json!({"expression": "1", "contextId": stale}), "first",
    ).await;
    assert!(stale_result.error.unwrap().message.contains("Cannot find context"));
    let other_result = cdp(
        &mut ctx, 5, "Runtime.evaluate",
        json!({"expression": "1", "contextId": other_page}), "second",
    ).await;
    assert!(other_result.error.is_none(), "other page context was pruned");
}
